src/client/clientmedia.h

```diff
@@ -38,7 +38,60 @@ struct HTTPFetchResult;
 bool clientMediaUpdateCache(const std::string &raw_hash,
 	const std::string &filedata);
 
-class ClientMediaDownloader
+// more of a base class than an interface but this name was most convenient...
+class IClientMediaDownloader
+{
+public:
+	virtual bool isStarted() const = 0;
+
+	// If this returns true, the downloader is done and can be deleted
+	virtual bool isDone() const = 0;
+
+	// Add a file to the list of required file (but don't fetch it yet)
+	virtual void addFile(const std::string &name, const std::string &sha1) = 0;
+
+	// Add a remote server to the list; ignored if not built with cURL
+	virtual void addRemoteServer(const std::string &baseurl) = 0;
+
+	// Steps the media downloader:
+	// - May load media into client by calling client->loadMedia()
+	// - May check media cache for files
+	// - May add files to media cache
+	// - May start remote transfers by calling httpfetch_async
+	// - May check for completion of current remote transfers
+	// - May start conventional transfers by calling client->request_media()
+	// - May inform server that all media has been loaded
+	//   by calling client->received_media()
+	// After step has been called once, don't call addFile/addRemoteServer.
+	virtual void step(Client *client) = 0;
+
+	// Must be called for each file received through TOCLIENT_MEDIA
+	// returns true if this file belongs to this downloader
+	virtual bool conventionalTransferDone(const std::string &name,
+			const std::string &data, Client *client) = 0;
+
+protected:
+	IClientMediaDownloader();
+	virtual ~IClientMediaDownloader() {};
+
+	// Forwards the call to the appropriate Client method
+	virtual bool loadMedia(Client *client, const std::string &data,
+		const std::string &name) = 0;
+
+	void createCacheDirs();
+
+	bool tryLoadFromCache(const std::string &name, const std::string &sha1,
+			Client *client);
+
+	bool checkAndLoad(const std::string &name, const std::string &sha1,
+			const std::string &data, bool is_from_cache, Client *client);
+
+	// Filesystem-based media cache
+	FileCache m_media_cache;
+	bool m_write_to_cache;
+};
+
+class ClientMediaDownloader : public IClientMediaDownloader
 {
```

Comment:
It would be good to disable class copy and move for this class btw.. The default copy and move cause evil stuff because of the destructor.